File: app/src/main/java/org/thoughtcrime/securesms/components/TransferControlView.java
package org.thoughtcrime.securesms.components;

import android.animation.LayoutTransition;
import android.content.Context;
import android.util.AttributeSet;
import android.view.View;
import android.widget.FrameLayout;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.core.content.ContextCompat;

import com.annimon.stream.Stream;
import com.pnikosis.materialishprogress.ProgressWheel;

import org.greenrobot.eventbus.EventBus;
import org.greenrobot.eventbus.Subscribe;
import org.greenrobot.eventbus.ThreadMode;
import org.thoughtcrime.securesms.R;
import org.thoughtcrime.securesms.attachments.Attachment;
import org.thoughtcrime.securesms.database.AttachmentDatabase;
import org.thoughtcrime.securesms.events.PartProgressEvent;
import org.thoughtcrime.securesms.mms.Slide;

import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public final class TransferControlView extends FrameLayout {

  private static final int UPLOAD_TASK_WEIGHT = 1;

  /**
   * A weighting compared to {@link #UPLOAD_TASK_WEIGHT}
   */
  private static final int COMPRESSION_TASK_WEIGHT = 3;

  @Nullable private List<Slide> slides;
  @Nullable private View        current;

  private final ProgressWheel progressWheel;
  private final View          downloadDetails;
  private final TextView      downloadDetailsText;

  private final Map<Attachment, Float> networkProgress;
  private final Map<Attachment, Float> compresssionProgress;

  public TransferControlView(Context context) {
    this(context, null);
  }

  public TransferControlView(Context context, AttributeSet attrs) {
    this(context, attrs, 0);
  }

  public TransferControlView(Context context, AttributeSet attrs, int defStyleAttr) {
    super(context, attrs, defStyleAttr);
    inflate(context, R.layout.transfer_controls_view, this);

    setLongClickable(false);
    setBackground(ContextCompat.getDrawable(context, R.drawable.transfer_controls_background));
    setVisibility(GONE);
    setLayoutTransition(new LayoutTransition());

    this.networkProgress      = new HashMap<>();
    this.compresssionProgress = new HashMap<>();

    this.progressWheel       = findViewById(R.id.progress_wheel);
    this.downloadDetails     = findViewById(R.id.download_details);
    this.downloadDetailsText = findViewById(R.id.download_details_text);
  }

  @Override
  public void setFocusable(boolean focusable) {
    super.setFocusable(focusable);
    downloadDetails.setFocusable(focusable);
  }

  @Override
  public void setClickable(boolean clickable) {
    super.setClickable(clickable);
    downloadDetails.setClickable(clickable);
  }

  @Override
  protected void onAttachedToWindow() {
    super.onAttachedToWindow();
    if (!EventBus.getDefault().isRegistered(this)) EventBus.getDefault().register(this);
  }

  @Override
  protected void onDetachedFromWindow() {
    super.onDetachedFromWindow();
    EventBus.getDefault().unregister(this);
  }

  public void setSlide(final @NonNull Slide slides) {
    setSlides(Collections.singletonList(slides));
  }

  public void setSlides(final @NonNull List<Slide> slides) {
    if (slides.isEmpty()) {
      throw new IllegalArgumentException("Must provide at least one slide.");
    }

    this.slides = slides;

    if (!isUpdateToExistingSet(slides)) {
      networkProgress.clear();
      compresssionProgress.clear();
      Stream.of(slides).forEach(s -> networkProgress.put(s.asAttachment(), 0f));
    }
    
    for (Slide slide : slides) {
      if (slide.asAttachment().getTransferState() == AttachmentDatabase.TRANSFER_PROGRESS_DONE) {
        networkProgress.put(slide.asAttachment(), 1f);
      }
    }

    switch (getTransferState(slides)) {
      case AttachmentDatabase.TRANSFER_PROGRESS_STARTED:
        showProgressSpinner(calculateProgress(networkProgress, compresssionProgress));
        break;
      case AttachmentDatabase.TRANSFER_PROGRESS_PENDING:
      case AttachmentDatabase.TRANSFER_PROGRESS_FAILED:
        downloadDetailsText.setText(getDownloadText(this.slides));
        display(downloadDetails);
        break;
      default:
        display(null);
        break;
    }
  }

  public void showProgressSpinner() {
    showProgressSpinner(calculateProgress(networkProgress, compresssionProgress));
  }

  public void showProgressSpinner(float progress) {
    if (progress == 0) {
      progressWheel.spin();
    } else {
      progressWheel.setInstantProgress(progress);
    }

    display(progressWheel);
  }

  public void setDownloadClickListener(final @Nullable OnClickListener listener) {
    downloadDetails.setOnClickListener(listener);
  }

  public void clear() {
    clearAnimation();
    setVisibility(GONE);
    if (current != null) {
      current.clearAnimation();
      current.setVisibility(GONE);
    }
    current = null;
    slides = null;
  }

  public void setShowDownloadText(boolean showDownloadText) {
    downloadDetailsText.setVisibility(showDownloadText ? VISIBLE : GONE);
    forceLayout();
  }

  private boolean isUpdateToExistingSet(@NonNull List<Slide> slides) {
    if (slides.size() != networkProgress.size()) {
      return false;
    }

    for (Slide slide : slides) {
      if (!networkProgress.containsKey(slide.asAttachment())) {
        return false;
      }
    }

    return true;
  }

  private int getTransferState(@NonNull List<Slide> slides) {
    int transferState = AttachmentDatabase.TRANSFER_PROGRESS_DONE;
    for (Slide slide : slides) {
      if (slide.getTransferState() == AttachmentDatabase.TRANSFER_PROGRESS_PENDING && transferState == AttachmentDatabase.TRANSFER_PROGRESS_DONE) {
        transferState = slide.getTransferState();
      } else {
        transferState = Math.max(transferState, slide.getTransferState());
      }
    }
    return transferState;
  }

  private String getDownloadText(@NonNull List<Slide> slides) {
    if (slides.size() == 1) {
      return slides.get(0).getContentDescription();
    } else {
      int downloadCount = Stream.of(slides).reduce(0, (count, slide) -> slide.getTransferState() != AttachmentDatabase.TRANSFER_PROGRESS_DONE ? count + 1 : count);
      return getContext().getResources().getQuantityString(R.plurals.TransferControlView_n_items, downloadCount, downloadCount);
    }
  }

  private void display(@Nullable final View view) {
    if (current == view) {
      return;
    }

    if (current != null) {
      current.setVisibility(GONE);
    }

    if (view != null) {
      view.setVisibility(VISIBLE);
      setVisibility(VISIBLE);
    } else {
      setVisibility(GONE);
    }

    current = view;
  }

  private static float calculateProgress(@NonNull Map<Attachment, Float> uploadDownloadProgress, Map<Attachment, Float> compresssionProgress) {
    float totalDownloadProgress    = 0;
    float totalCompressionProgress = 0;

    for (float progress : uploadDownloadProgress.values()) {
      totalDownloadProgress += progress;
    }

    for (float progress : compresssionProgress.values()) {
      totalCompressionProgress += progress;
    }

    float weightedProgress = UPLOAD_TASK_WEIGHT * totalDownloadProgress         + COMPRESSION_TASK_WEIGHT * totalCompressionProgress;
    float weightedTotal    = UPLOAD_TASK_WEIGHT * uploadDownloadProgress.size() + COMPRESSION_TASK_WEIGHT * compresssionProgress.size();

    return weightedProgress / weightedTotal;
  }

  @Subscribe(sticky = true, threadMode = ThreadMode.MAIN)
  public void onEventAsync(final PartProgressEvent event) {
    if (networkProgress.containsKey(event.attachment)) {
      float proportionCompleted = ((float) event.progress) / event.total;

      if (event.type == PartProgressEvent.Type.COMPRESSION) {
        compresssionProgress.put(event.attachment, proportionCompleted);
      } else {
        networkProgress.put(event.attachment, proportionCompleted);
      }

      progressWheel.setInstantProgress(calculateProgress(networkProgress, compresssionProgress));
    }
  }
}

File: app/src/main/java/org/thoughtcrime/securesms/components/transfercontrols/TransferControlView.kt
/*
 * Copyright 2023 Signal Messenger, LLC
 * SPDX-License-Identifier: AGPL-3.0-only
 */
package org.thoughtcrime.securesms.components.transfercontrols

import android.content.Context
import android.util.AttributeSet
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import androidx.compose.ui.platform.AbstractComposeView
import androidx.compose.ui.platform.ViewCompositionStrategy
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.cancel
import kotlinx.coroutines.launch
import org.greenrobot.eventbus.EventBus
import org.greenrobot.eventbus.Subscribe
import org.greenrobot.eventbus.ThreadMode
import org.signal.core.models.database.AttachmentId
import org.signal.core.ui.compose.theme.SignalTheme
import org.signal.core.util.ByteSize
import org.signal.core.util.ThrottledDebouncer
import org.signal.core.util.bytes
import org.signal.core.util.logging.Log
import org.thoughtcrime.securesms.attachments.Attachment
import org.thoughtcrime.securesms.attachments.DatabaseAttachment
import org.thoughtcrime.securesms.components.RecyclerViewParentTransitionController
import org.thoughtcrime.securesms.database.AttachmentTable
import org.thoughtcrime.securesms.events.PartProgressEvent
import org.thoughtcrime.securesms.jobs.AttachmentBackfill
import org.thoughtcrime.securesms.mms.Slide
import org.thoughtcrime.securesms.util.MediaUtil
import java.util.UUID

/**
 * Displays the start/cancel/progress controls that overlay an attachment thumbnail.
 */
class TransferControlView @JvmOverloads constructor(context: Context, attrs: AttributeSet? = null, defStyleAttr: Int = 0) : AbstractComposeView(context, attrs, defStyleAttr) {

  companion object {
    private val TAG = Log.tag(TransferControlView::class.java)

    /** Flip to true locally to trace a single view's render transitions and ignored progress events. */
    private const val VERBOSE_DEVELOPMENT_LOGGING = false
  }

  private var state = TransferControlViewState()

  /** Throttled observable flow of [state] */
  private var renderState by mutableStateOf<TransferControlsRenderState>(TransferControlsRenderState.Gone)

  private val progressUpdateDebouncer = ThrottledDebouncer(100)

  private var previousRenderState: TransferControlsRenderState = TransferControlsRenderState.Gone

  /** Active while view is attached */
  private var renderScope: CoroutineScope? = null

  /** Per-instance id so a single recycled view can be isolated in logcat when [VERBOSE_DEVELOPMENT_LOGGING] is on. */
  private val viewId by lazy { UUID.randomUUID().toString().take(8) }

  init {
    setViewCompositionStrategy(ViewCompositionStrategy.DisposeOnDetachedFromWindowOrReleasedFromPool)
    isLongClickable = false
    addOnAttachStateChangeListener(RecyclerViewParentTransitionController(child = this))
  }

  @Composable
  override fun Content() {
    SignalTheme {
      TransferControls(
        state = renderState,
        onStartClick = { state.startTransferClickListener?.onClick(this) },
        onCancelClick = { state.cancelTransferClickedListener?.onClick(this) },
        onPlayClick = { state.instantPlaybackClickListener?.onClick(this) }
      )
    }
  }

  override fun onAttachedToWindow() {
    super.onAttachedToWindow()
    if (!EventBus.getDefault().isRegistered(this)) EventBus.getDefault().register(this)

    renderScope = CoroutineScope(Dispatchers.Main.immediate).also { scope ->
      scope.launch {
        AttachmentBackfill.awaiting.collect { renderState() }
      }
    }
  }

  override fun onDetachedFromWindow() {
    super.onDetachedFromWindow()
    EventBus.getDefault().unregister(this)
    renderScope?.cancel()
    renderScope = null
  }

  fun isGone(): Boolean {
    val state = TransferControls.deriveRenderState(state = state, awaitingAttachmentIds = state.slides.attachmentIdsAwaitingBackfill())
    return state is TransferControlsRenderState.Gone
  }

  private fun updateState(stateFactory: (TransferControlViewState) -> TransferControlViewState) {
    state = stateFactory(state)
    renderState()
  }

  /**
   * Derives and applies the render state by combining [state] with the current awaiting attachments. Both the state and
   * awaiting paths change events funnel through here.
   */
  private fun renderState() {
    val oldRenderState = previousRenderState
    val newRenderState = TransferControls.deriveRenderState(state, state.slides.attachmentIdsAwaitingBackfill())
    previousRenderState = newRenderState

    if (oldRenderState == newRenderState) {
      return
    }

    verboseLog { "render $oldRenderState -> $newRenderState slides=[${slidesAsLogString(state.slides)}]" }

    // Only throttle noisy progress changes
    if (oldRenderState is TransferControlsRenderState.InProgress && oldRenderState.isProgressOnlyDifference(newRenderState)) {
      progressUpdateDebouncer.publish {
        renderState = newRenderState
        visibility = VISIBLE
      }
    } else {
      progressUpdateDebouncer.clear()
      renderState = newRenderState
      if (newRenderState !is TransferControlsRenderState.Gone) {
        visibility = VISIBLE
      }
    }
  }

  @Subscribe(sticky = true, threadMode = ThreadMode.MAIN)
  fun onEventAsync(event: PartProgressEvent) {
    val attachment = event.attachment
    updateState {
      if (!it.networkProgress.containsKey(attachment)) {
        verboseLog { "Ignoring progress event for an attachment not in this view's slide set (likely a recycled view). ts=${attachment.uploadTimestamp}" }
        return@updateState it
      }

      if (event.type == PartProgressEvent.Type.COMPRESSION) {
        val progress = it.compressionProgress.toMutableMap()
        progress.applyProgress(attachment, Progress.fromEvent(event))
        return@updateState it.copy(compressionProgress = progress.toMap())
      } else {
        val progress = it.networkProgress.toMutableMap()
        progress.applyProgress(attachment, Progress.fromEvent(event))
        return@updateState it.copy(networkProgress = progress.toMap())
      }
    }
  }

  fun setSlides(slides: List<Slide>) {
    require(slides.isNotEmpty()) { "Must provide at least one slide." }
    clearResolvedBackfills(slides)
    updateState { state ->
      val isNewSlideSet = !isUpdateToExistingSet(state, slides)
      val networkProgress: MutableMap<Attachment, Progress> = if (isNewSlideSet) HashMap() else state.networkProgress.toMutableMap()
      if (isNewSlideSet) {
        slides.forEach { networkProgress[it.asAttachment()] = Progress(0.bytes, it.fileSize.bytes) }
      }
      val compressionProgress: MutableMap<Attachment, Progress> = if (isNewSlideSet) HashMap() else state.compressionProgress.toMutableMap()
      var allStreamableOrDone = true
      for (slide in slides) {
        val attachment = slide.asAttachment()
        if (attachment.transferState == AttachmentTable.TRANSFER_PROGRESS_DONE) {
          networkProgress[attachment] = Progress(attachment.size.bytes, attachment.size.bytes)
        } else if (!MediaUtil.isInstantVideoSupported(slide)) {
          allStreamableOrDone = false
        }
      }
      val playableWhileDownloading = allStreamableOrDone
      val isUpload = slides.all {
        (it.asAttachment() as? DatabaseAttachment)?.hasData == true
      }

      state.copy(
        slides = slides,
        networkProgress = networkProgress,
        compressionProgress = compressionProgress,
        playableWhileDownloading = playableWhileDownloading,
        isUpload = isUpload
      )
    }
  }

  private fun isUpdateToExistingSet(currentState: TransferControlViewState, slides: List<Slide>): Boolean {
    if (slides.size != currentState.networkProgress.size) {
      return false
    }
    for (slide in slides) {
      if (!currentState.networkProgress.containsKey(slide.asAttachment())) {
        return false
      }
    }
    return true
  }

  fun setTransferClickListener(listener: OnClickListener) {
    updateState { it.copy(startTransferClickListener = listener) }
  }

  fun setCancelClickListener(listener: OnClickListener) {
    updateState { it.copy(cancelTransferClickedListener = listener) }
  }

  fun setInstantPlaybackClickListener(listener: OnClickListener) {
    updateState { it.copy(instantPlaybackClickListener = listener) }
  }

  fun clear() {
    visibility = GONE
    updateState { TransferControlViewState() }
  }

  fun setShowSecondaryText(showSecondaryText: Boolean) {
    updateState { it.copy(showSecondaryText = showSecondaryText) }
  }

  fun setVisible(isVisible: Boolean) {
    updateState { it.copy(isVisible = isVisible) }
  }

  override fun setFocusable(focusable: Boolean) {
    super.setFocusable(false)
    updateState { it.copy(isFocusable = focusable) }
  }

  override fun setClickable(clickable: Boolean) {
    super.setClickable(false)
    updateState { it.copy(isClickable = clickable) }
  }

  private fun List<Slide>.attachmentIdsAwaitingBackfill(): Set<AttachmentId> {
    return this
      .mapNotNullTo(HashSet()) { (it.asAttachment() as? DatabaseAttachment)?.attachmentId }
      .apply { retainAll(AttachmentBackfill.awaiting.value) }
  }

  /** Tells [AttachmentBackfill] to stop awaiting any backfilled attachment that this view now sees as DONE. */
  private fun clearResolvedBackfills(slides: List<Slide>) {
    for (slide in slides) {
      val attachment = slide.asAttachment() as? DatabaseAttachment ?: continue
      if (attachment.transferState == AttachmentTable.TRANSFER_PROGRESS_DONE && attachment.attachmentId in AttachmentBackfill.awaiting.value) {
        AttachmentBackfill.onAttachmentTerminal(attachment.attachmentId, attachment.mmsId)
      }
    }
  }

  private fun MutableMap<Attachment, Progress>.applyProgress(attachment: Attachment, update: Progress) {
    if (update.completed < 0.bytes) {
      remove(attachment)
    } else {
      put(attachment, update)
    }
  }

  private inline fun verboseLog(message: () -> String) {
    if (VERBOSE_DEVELOPMENT_LOGGING) {
      Log.d(TAG, "[$viewId] ${message()}")
    }
  }

  private fun slidesAsLogString(slides: List<Slide>): String {
    return slides.joinToString { "ts=${it.asAttachment().uploadTimestamp},xfer=${it.transferState}" }
  }

  data class Progress(val completed: ByteSize, val total: ByteSize) {
    companion object {
      fun fromEvent(event: PartProgressEvent): Progress {
        return Progress(event.progress.bytes, event.total.bytes)
      }
    }
  }
}
